sites/funmooc/src/backend/base/context_processors.py

```diff
@@ -0,0 +1,63 @@
+"""
+Template context_processors
+"""
+import json
+
+from django.conf import settings
+from django.utils.translation import get_language_from_request
+
+from richie.apps.courses.models import Organization
+
+
+def marketing_metas(request):
+    """
+    Context processor to add all information required by Marketing frontend scripts.
+    """
+
+    page = request.current_page or None
+    language = get_language_from_request(request, check_path=True)
+
+    def get_organizations():
+        """
+        Return all organizations linked to the current page via an organization plugin
+        in any of the placeholders on the page.
+        """
+        organizations = []
+
+        # We need to be aware if an organisation is linked to the current page.
+        # This is the case for the organization_detail page where
+        # there is no organization plugin but the page has an organization attribute.
+        if hasattr(page, "organization"):
+            organizations.append(page.organization.code)
+
+        selector = "extended_object__organization_plugins__cmsplugin_ptr"
+        filter_dict = {
+            "{:s}__language".format(selector): language,
+            "{:s}__placeholder__page".format(selector): page,
+            "extended_object__title_set__published": True,
+        }
+
+        organizations += list(
+            organization.code
+            for organization in Organization.objects.filter(**filter_dict)
+            .select_related("extended_object")
+            .distinct()
+        )
+
+        return organizations
+
+    context = {
+        "MARKETING_CONTEXT": json.dumps(
+            {
+                "xiti": {
+                    "level2": str(page.node.get_root().pk),
+                    "organizations": get_organizations() if not page.is_home else [],
```

Comment:
I'm wondering if these queries will be done on each page load or if they will be skipped when the page is cached. Probably yes but have you checked it?

Reply:
I have to write some tests for this context_processor